class/class.midtrans-gateway-subscription.php

```diff
@@ -0,0 +1,225 @@
+      $params['enabled_payments'] = ['credit_card'];
+      // add bank & channel migs params.
+      if (strlen($this->get_option('acquring_bank')) > 0)
+        $params['credit_card']['bank'] = strtoupper ($this->get_option('acquring_bank'));
+      // add bin params        
+      if (strlen($this->get_option('bin_number')) > 0){
+        $bins = explode(',', $this->get_option('bin_number'));
+        $params['credit_card']['whitelist_bins'] = $bins;
+      }
+
+      // Empty the cart because payment is initiated.
+      $woocommerce->cart->empty_cart();
+      try {
+        $snapResponse = WC_Midtrans_API::createSnapTransaction( $params, $this->id );
+      } catch (Exception $e) {
+        $this->setLogError( $e->getMessage() );
+        WC_Midtrans_Utils::json_print_exception( $e, $this );
+        exit();
+      }
+
+      if(property_exists($this,'enable_redirect') && $this->enable_redirect == 'yes'){
+        $redirectUrl = $snapResponse->redirect_url;
+      }else{
+        $redirectUrl = $order->get_checkout_payment_url( true )."&snap_token=".$snapResponse->token;
+      }
+
+      // Add snap token & snap redirect url to $order metadata.
+      $order->update_meta_data('_mt_payment_snap_token',$snapResponse->token);
+      $order->update_meta_data('_mt_payment_url',$snapResponse->redirect_url);
+      $order->save();
+      
+      if(property_exists($this,'enable_immediate_reduce_stock') && $this->enable_immediate_reduce_stock == 'yes'){
+        wc_reduce_stock_levels($order);
+      }
+
+      $successResponse['redirect'] = $redirectUrl;
+      return $successResponse;
+    }
+
+    /**
+     * Hook function that will be called on receipt page.
+     * Output HTML for Snap payment page. Including `snap.pay()` part.
+     * @param  string $order_id generated by WC.
+     * @return string HTML.
+     */
+    function receipt_page( $order_id ) {
+      global $woocommerce;
+      $pluginName = 'subscription';
+      require_once(dirname(__FILE__) . '/payment-page.php'); 
+    }
+
+    /**
+     * Scheduled_subscription_payment function.
+     * This function is called when renewal order is triggered.
+     * 
+     * @param float $amount_to_charge float The amount to charge.
+     * @param WC_Order $renewal_order A WC_Order object created to record the renewal payment.
+     */
+    public function scheduled_subscription_payment( $amount_to_charge, $renewal_order ) {
+      if (!$renewal_order) {  
+        return array ( 'result' => 'failure' );
+      }
+      $order_id = $renewal_order->get_id();
+      $subscription_order = wcs_get_subscriptions_for_renewal_order( $order_id );
+      $checkout_payment_url = $renewal_order->get_checkout_payment_url();
+
+      // Retrieve card token from meta.
+      foreach ( $subscription_order as $subscription ) {
+        $card_token = $subscription->get_meta('_mt_subscription_card_token');
+      }
+
+      // if card_token null, the transaction will fail.
+      if ($card_token == '' ) {
```

Comment:
yes if the get_meta() failed, it will return an empty string
i think need to replace the comment to `// if can't find the card token, the transaction will fail`

Reply:
Yes. Ok cool.

Reply:
done on https://github.com/veritrans/SNAP-Woocommerce/pull/39/commits/cd34815e09c46b3c4024b641b9f1d2606118dc2e